Support for other webcams than /dev/video0, fixed a regular expression

bionik opened the pull request:
This patch tries to allow detection of usb webcams. It also corrects a small regular expression problem.
* When polling for webcams, we check for cameras by listing all video devices, and if at least one is found, we give the path of the first device to startUsb function. 
* StartUsb will use the device given as the parameter. It also has an option override the first found device which is pssed as a parameter, if it's explicitly set in the /boot/octopi.txt.
* If the device is not given it the octopi.txt, we will add the device path as a parameter for mjpeg-streamer, into the options variable.
* The unescaped "+" sign that was used in the regular expression with sed, to get data for the extracted_device variable, is now also correctly escaped, and the expression works now on OctoPi.
* The uevent check will now also use the correct device identifier, because the regular expression now parses the device set in octopi.txt correctly.

Tests:
* I have tested the device override; the /boot/octopi.txt will override the automatically detected devices. The device in octopi.txt is passed to mjpeg-streamer & uevent check correctly.
* I have tested that when plugging many usb video devices and unplugging them so that the device id changes, the script will now find /dev/video1 by itself and spawn mjpeg-streamer with that device correctly (when the explicit device is not given in /boot/octopi.txt)
* Boots fine and the stream gets up after boot
* Regular expression fix: (used in /boot/octopi.txt device override / extracted_device)

Incorrect behaviour, before the fix:

```
  pi@octopi:~ $ echo `echo "-r 1920x1080 -d /dev/video32 -e -t pal" | sed 's@.*-d /dev/\(video[0-9]+\).*@\1@'`
  -r 1920x1080 -d /dev/video32 -e -t pal
  pi@octopi:~ $ echo `echo "-r 1920x1080 -d /dev/video0 -e -t pal" | sed 's@.*-d /dev/\(video[0-9]+\).*@\1@'`
  -r 1920x1080 -d /dev/video0 -e -t pal
```

Correct behaviour on the Raspberry Pi after the fix:

```
  pi@octopi:~ $ echo `echo "-r 1920x1080 -d /dev/video3 -e -t pal" | sed 's@.*-d /dev/\(video[0-9]\+\).*@\1@'`
  video3
  pi@octopi:~ $ echo `echo "-r 1920x1080 -d /dev/video32 -e -t pal" | sed 's@.*-d /dev/\(video[0-9]\+\).*@\1@'`
  video32
```

guysoft commented:
@foosel I am not really following to the letter and would appreciate if you review it

bionik commented:
I was thinking about the case you mentioned earlier, and I thought that the hard coded device in /boot/octopi.txt is literally hard coded and should be respected, even though it would not necessarily work.
I am happy to modify the code like you suggested, it seems like "the correct way". Thank you for your input!

foosel commented:
I agree that it should be respected, but I think forcefully going forward even if we can already know that things will fail would be wrong, hence my comment :)

bionik commented:
Agreed! Your suggestion seems great, I will implement it when I get off from work :)

bionik commented:
I modified the code like @foosel suggested:
* If device is detected in options, it is parsed and removed from the options string
* When performing device detection, we will check if device is set / detected in the options
  * If device is set explicitly, we will check if that device is connected. If it is, streamer is invoked with that device. If not, we will complain about it and try to detect the camera again in two minutes.
  * If device is not explicitly set, we will invoke the streamer with the first found camera.
* In startUsb we add the device which was given during invoking back into the options and check that device against bad cameras

I have tested again on OctoPi, with two webcameras. I tried with and without explicitly setting the cameras, with incorrect camera device.

I have added some logging / echoing to explain better what is going on, happy to remove them if you guys think it's too spammy.

foosel commented:
> I have added some logging / echoing to explain better what is going on, happy to remove them if you guys think it's too spammy.

In general no, but I think logging "Found video devices" even if none are detected is a bit weird. I think it would be better to quickly move that line down into the if block testing if at least one device was detected and we are even supposed to trigger on usb devices.

Otherwise this looks great now! 👍

bionik commented:
Good suggestion, as usual :) Thank you for your help and support!

foosel commented:
From my point of view nothing speaks against a merge now, the rest is up to @guysoft ;)
